routing_plugin.go

```diff
@@ -104,3 +120,419 @@ func ReleaseRoutingPlugin(ctx context.Context) {
+		AmtMsat:           lnwire.MilliSatoshi(r.amount * 1000),
+		FeeLimitMsat:      lnwire.MilliSatoshi(r.amount * 1000),
+		UseMissionControl: true,
+	}
+
+	if log.Level() == btclog.LevelTrace {
+		res, err := r.lnd.Client.QueryRoutes(ctx, queryRoutesReq)
+		log.Tracef("BeforePayment() QueryRoutes(1)=%v, err=%v",
+			spew.Sdump(res), err)
+	}
+
+	// Do not do anything unless we tried to route the payment at least
+	// once.
+	if currAttempt < 2 {
+		return nil
+	}
+
+	// Calculate the limit until we'll disable edges. Anything above this
+	// limit should be encouraged.
+	minFee := r.nodesByMaxFee[0].fee
+	maxFee := r.nodesByMaxFee[len(r.nodesByMaxFee)-1].fee
+	limit := minFee + ((maxFee-minFee)/int64(maxAttempts))*int64(currAttempt)
+
+	// Create a timestamp just slightly in the future as Mission Control
+	// stores timestamps with sub second precision where as we send a unix
+	// timestamp it may occur that we can't override the last entries as
+	// they have the same unix timestamp.
+	// TODO(bhandras): not very reliable, ideally we'd need a force import
+	// for MC.
+	now := r.clock.Now().Add(time.Second)
+
+	allowed := 0
+	var entries []lndclient.MissionControlEntry
+	for _, nodeFeeInfo := range r.nodesByMaxFee {
+		if nodeFeeInfo.fee < limit {
+			log.Debugf("Discouraging payments from %v to %v",
+				nodeFeeInfo.node, r.target)
+			entries = append(
+				entries, lndclient.MissionControlEntry{
+					NodeFrom: nodeFeeInfo.node,
+					NodeTo:   r.target,
+					FailTime: now,
+					FailAmt:  1,
+				})
+		} else {
+			log.Debugf("Encouraging payments from %v to %v",
+				nodeFeeInfo.node, r.target)
+			entries = append(
+				entries, lndclient.MissionControlEntry{
+					NodeFrom:    nodeFeeInfo.node,
+					NodeTo:      r.target,
+					SuccessTime: now,
+					SuccessAmt: lnwire.MilliSatoshi(
+						nodeFeeInfo.capacity * 1000,
+					),
+				})
+			allowed++
+		}
+	}
+
+	// There's no point retrying the payment since we discouraged using
+	// all inbound peers to the target.
+	if allowed == 0 {
+		return ErrRoutingPluginNoMoreRetries
+	}
+
+	err := r.lnd.Router.ImportMissionControl(ctx, entries)
+	log.Tracef("Imported MC state: %v, err=%v",
+		spew.Sdump(entries), err)
+
+	if err == nil && log.Level() == btclog.LevelTrace {
+		res, err := r.lnd.Client.QueryRoutes(ctx, queryRoutesReq)
+		log.Tracef("BeforePayment() QueryRoutes(2)=%v, err=%v",
+			spew.Sdump(res), err)
+	}
+
+	return err
+}
+
+// Done will reset the mission control state to where it was before using the
+// routing plugin.
+func (r *lowToHighRoutingPlugin) Done(ctx context.Context) error {
+	if r.mcState == nil {
+		return nil
+	}
+
+	var err error
+	if len(r.mcState) > 0 {
+		now := r.clock.Now()
+		// Roll the entry times forward (to be able to override recent
+		// updates). TODO: maybe we should reset here first instead.
+		for i := range r.mcState {
+			if !r.mcState[i].FailTime.IsZero() {
+				r.mcState[i].FailTime = now
+			}
+
+			if !r.mcState[i].SuccessTime.IsZero() {
```

Comment:
Do we need to roll our whole mission control state forward here? Could be a less invasive approach to keep track of the pairs that we've modified and only reset them to their previous state (+ rolled forward time).

Motivation here is that unaffected results will still be "cooled down" based on their timestamp, rather than being brought forward to the present (which could make routing worse even? although idk)

Reply:
Yes thought about this myself too. Actually waited for some feedback from reviewers before refining what we store and how we restore.

Reply:
We've got our full mission control state in `r.mcstate`, yes? 

Since the idea here was to minimally disrupt mission control, I think that a good reset option would be to track the edges that we actually modified in `BeforePayment` and only reset those by rolling them forward.